staging/src/k8s.io/apiserver/pkg/util/version/version.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package version
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/spf13/pflag"
+	"k8s.io/apimachinery/pkg/util/version"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	"k8s.io/component-base/featuregate"
+	baseversion "k8s.io/component-base/version"
+)
+
+var DefaultEffectiveVersionRegistry EffectiveVersionRegistry = NewEffectiveVersionRegistry()
+
+const (
+	ComponentGenericAPIServer = "k8s.io/apiserver"
+)
+
+type EffectiveVersionRegistry interface {
+	// EffectiveVersionFor returns the EffectiveVersion registered under the component.
+	// Returns nil if the component is not registered.
+	EffectiveVersionFor(component string) MutableEffectiveVersion
+	// EffectiveVersionForOrDefault returns the EffectiveVersion registered under the component.
+	// Returns the default EffectiveVersion if the component is not registered, and registers the default EffectiveVersion under component.
+	EffectiveVersionForOrDefault(component string) MutableEffectiveVersion
+	// RegisterEffectiveVersionFor registers the EffectiveVersion for a component.
+	// Overrides existing EffectiveVersion if it is already in the registry.
+	RegisterEffectiveVersionFor(component string, ver MutableEffectiveVersion)
+}
+
+type effectiveVersionRegistry struct {
+	effectiveVersions map[string]MutableEffectiveVersion
+	mutex             sync.RWMutex
+}
+
+func NewEffectiveVersionRegistry() EffectiveVersionRegistry {
+	return &effectiveVersionRegistry{effectiveVersions: map[string]MutableEffectiveVersion{}}
+}
+
+func (r *effectiveVersionRegistry) EffectiveVersionFor(component string) MutableEffectiveVersion {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	return r.effectiveVersions[component]
+}
+
+func (r *effectiveVersionRegistry) EffectiveVersionForOrDefault(component string) MutableEffectiveVersion {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	ver, ok := r.effectiveVersions[component]
+	if ok {
+		return ver
+	}
+	ver = NewEffectiveVersion(baseversion.Get().String())
+	r.effectiveVersions[component] = ver
```

Comment:
I wasn't expecting this function to mutate the effectiveVersions map and change where subsequent `EffectiveVersionFor` calls might return. (If we do make this read-only we can also use RLock here).

Reply:
Changed the name to be `EffectiveVersionForOrRegister`. I find this useful because it is the shorthand for the steps to always to have to check `EffectiveVersionFor` is nil and register if nil

Reply:
No, we do not check if the version passed to this function does not match what is returned. The idea is if that matters, the user should probably call `EffectiveVersionFor` and `RegisterEffectiveVersionFor` directly.

Reply:
Ok, so long as this is clearly named/documented as a check-and-set type operation, SGTM.